src/emu/emuopts.cpp

```diff
@@ -213,133 +218,777 @@ const options_entry emu_options::s_option_entries[] =
+								// for which we have not already distributed a part to
+								device_image_interface *image = software_list_device::find_mountable_image(
+									config,
+									swpart,
+									[&results](const device_image_interface &candidate) { return results.image.count(candidate.instance_name()) == 0; });
+
+								// did we find a slot to put this part into?
+								if (image != nullptr)
+								{
+									// we've resolved this software
+									results.image[image->instance_name()] = string_format("%s:%s:%s", swlistdev.list_name(), software_name, swpart.name());
+
+									// does this software part have a requirement on another part?
+									const char *requirement = swpart.feature("requirement");
+									if (requirement)
+										software_identifier_stack.push(requirement);
+								}
+								compatible = true;
+							}
+							found = true;
+						}
+
+						// identify other shared features specified as '<<slot name>>_default'
+						//
+						// example from SMS:
+						//
+						//  <software name = "alexbmx">
+						//      ...
+						//      <sharedfeat name = "ctrl1_default" value = "paddle" />
+						//  </software>
+						for (const feature_list_item &fi : swinfo->shared_info())
+						{
+							const std::string default_suffix = "_default";
+							if (fi.name().size() > default_suffix.size()
+								&& fi.name().compare(fi.name().size() - default_suffix.size(), default_suffix.size(), default_suffix) == 0)
+							{
+								std::string slot_name = fi.name().substr(0, fi.name().size() - default_suffix.size());
+								results.slot[slot_name] = fi.value();
+							}
+						}
+					}
+				}
+				if (compatible)
+					break;
+			}
+
+			if (!compatible)
+			{
+				software_list_device::display_matches(config, nullptr, software_name);
+				if (!found)
+					throw options_exception(options_exception::condition_type::ERROR, "");
+				else
+					throw options_exception(options_exception::condition_type::ERROR, "Software '%s' is incompatible with system '%s'\n", software_name, m_system->name);
+			}
+		}
 	}
+	return results;
+}
+
+
+//-------------------------------------------------
+//	slot_option
+//-------------------------------------------------
+
+const slot_option &emu_options::slot_option(const std::string &device_name) const
+{
+	auto iter = m_slot_options.find(device_name);
+	assert(iter != m_slot_options.end() && "Attempt to access non-existent slot option");
+	return iter->second;
+}
+
+slot_option &emu_options::slot_option(const std::string &device_name)
+{
+	auto iter = m_slot_options.find(device_name);
+	assert(iter != m_slot_options.end() && "Attempt to access non-existent slot option");
+	return iter->second;
+}
+
 
-	return false;
+//-------------------------------------------------
+//	has_slot_option
+//-------------------------------------------------
+
+bool emu_options::has_slot_option(const std::string &device_name) const
+{
+	return m_slot_options.count(device_name) > 0;
```

Comment:
`x.count(foo)` can be substantially slower than `x.find(foo) != x.end()` if you don't care about the actual number of matches.